### DISCUSSION TITLE
How can I guarantee contract requirements?

### DISCUSSION THREAD

**Discussion by carpnick** [aws-cloudformation/cloudformation-cli-python-plugin#243] [Q&A]
Hi folks,

Looking at the [contract](https://docs.aws.amazon.com/cloudformation-cli/latest/userguide/resource-type-test-contract.html) as written, how do I guarantee the following lines?(see quoted below). From what I can tell, no plugin(this one included) contains the original start time of the resource call.  Is it my handler's problem to  track original call time in the callback context?  Shouldnt this be included on the generated classes or within the framework Models?  I can then just interrogate and respond with error, or even better let framework respond with error...  Right now as written in contract - we dont know when to return since we do not have the start time and the resource schema of timeoutinMinutes is not exposed to us via any generated classes, correct?  I see it is used to set maximum time on assuming the role only.  That doesnt allow us to honor contract for less than 60 minutes for a single request: see [here](https://github.com/aws-cloudformation/cloudformation-cli/blob/master/src/rpdk/core/project.py#L511-L522)

Timeout Implementation:
~~- Do we have to guarantee any kind of runtime for functions?  IE if this is running in Lambda is it our responsibility to kill function before hitting 15 minutes?  Shouldn't it be the framework's job to give us total allowable execution time?  Nothing in contract docs state allowable execution time.  Ideally you would include a line like : `A list/create/update/delete/read request MUST return within NN seconds of continuous execution.  If you need more time, add data to callback context and handle in next call`~~

^ is handled by `MUST always return a ProgressEvent object within 60 seconds`

**Contract Lines**
```
A create handler MUST return FAILED progress event if it can't reach the desired-state within the timeout specified in the resource schema.

An update handler MUST return FAILED progress event if it can't reach the desired-state within the timeout specified in the resource schema.

A delete handler MUST return FAILED progress event if it can't reach the desired-state within the timeout specified in the resource schema.

```

**Comment by carpnick** [comment]
Any help would be appreciated here....?

**Comment by ericzbeard** [comment]
You do have to track this yourself on the callback context. You're right that it would be easier if we tracked this automatically.

**Reply by carpnick** [reply]
I guess this is a feature request :).

Should this stay here or should I move it to an issue?  Does it belong in this repository or the CLI repository?

**Reply by ericzbeard** [reply]
Should be an issue in this repo.

**Reply by carpnick** [reply]
Issue opened - https://github.com/aws-cloudformation/cloudformation-cli-python-plugin/issues/245
